pkg/executioncluster/impl/random_cluster_selector.go

```diff
@@ -0,0 +1,173 @@
+		}
+		executionTargetMap[cluster.Name] = *executionTarget
+	}
+	return executionTargetMap, nil
+}
+
+func getDomainsForCluster(cluster runtime.ClusterConfig, domainMap map[string]runtime.Domain) ([]string, error) {
+	if len(cluster.AllowedDomains) == 0 {
+		allDomains := make([]string, len(domainMap))
+		index := 0
+		for id := range domainMap {
+			allDomains[index] = id
+			index++
+		}
+		return allDomains, nil
+	}
+	for _, allowedDomain := range cluster.AllowedDomains {
+		if _, ok := domainMap[allowedDomain]; !ok {
+			return nil, fmt.Errorf("invalid domain %s", allowedDomain)
+		}
+	}
+	return cluster.AllowedDomains, nil
+}
+
+func getDomainWeightedRandomForCluster(ctx context.Context, scope promutils.Scope, executionTargetProvider interfaces.ExecutionTargetProvider,
+	clusterConfig runtime.ClusterConfiguration,
+	domainMap map[string]runtime.Domain) (map[string]random.WeightedRandomList, error) {
+	domainEntriesMap := make(map[string][]random.Entry)
+	for _, cluster := range clusterConfig.GetClusterConfigs() {
+		// If cluster is not enabled, it is not eligible for selection
+		if !cluster.Enabled {
+			continue
+		}
+		executionTarget, err := executionTargetProvider.GetExecutionTarget(scope, cluster)
+		if err != nil {
+			return nil, err
+		}
+		targetEntry := random.Entry{
+			Item:   *executionTarget,
+			Weight: cluster.Weight,
+		}
+		clusterDomains, err := getDomainsForCluster(cluster, domainMap)
+		if err != nil {
+			return nil, err
+		}
+		for _, domain := range clusterDomains {
+			if _, ok := domainEntriesMap[domain]; ok {
+				domainEntriesMap[domain] = append(domainEntriesMap[domain], targetEntry)
+			} else {
+				domainEntriesMap[domain] = []random.Entry{targetEntry}
+			}
+		}
+	}
+	domainWeightedRandomMap := make(map[string]random.WeightedRandomList)
+	for domain, entries := range domainEntriesMap {
+		weightedRandomList, err := random.NewWeightedRandom(ctx, entries)
+		if err != nil {
+			return nil, err
+		}
+		domainWeightedRandomMap[domain] = weightedRandomList
+	}
+	return domainWeightedRandomMap, nil
+}
+
+func (s RandomClusterSelector) GetAllValidTargets() []executioncluster.ExecutionTarget {
+	v := make([]executioncluster.ExecutionTarget, 0)
+	for _, value := range s.executionTargetMap {
+		if value.Enabled {
+			v = append(v, value)
+		}
+	}
+	return v
+}
+
+func (s RandomClusterSelector) GetTarget(spec *executioncluster.ExecutionTargetSpec) (*executioncluster.ExecutionTarget, error) {
+	if spec == nil || (spec.TargetID == "" && spec.ExecutionID == nil) {
+		return nil, fmt.Errorf("invalid executionTargetSpec %v", spec)
+	}
+	if spec.TargetID != "" {
+		if val, ok := s.executionTargetMap[spec.TargetID]; ok {
+			return &val, nil
+		}
+		return nil, fmt.Errorf("invalid cluster target %s", spec.TargetID)
+	}
+	if spec.ExecutionID != nil {
+		if weightedRandomList, ok := s.domainWeightedRandomMap[spec.ExecutionID.GetDomain()]; ok {
+			executionName := spec.ExecutionID.GetName()
+			if executionName != "" {
```

Comment:
Is it legit to have an execution ID with no name?

Reply:
Not in general, but someone can use `random_cluster_selector.go` without passing in executionID